src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,511 @@
+			dead_oids = lappend_oid(dead_oids, current_entry->db_oid);
+	}
+
+	foreach(lc, dead_oids)
+	{
+		Oid			dead_oid = lfirst_oid(lc);
+
+		tsbgw_total_workers_decrement();
+		hash_search(db_htab, &dead_oid, HASH_REMOVE, &found);
+		Assert(found);
+	}
+	list_free(dead_oids);
+}
+
+
+static void
+stop_all_db_schedulers(HTAB *db_htab)
+{
+	HASH_SEQ_STATUS hash_seq;
+	tsbgwHashEntry *current_entry;
+
+	hash_seq_init(&hash_seq, db_htab);
+
+
+
+	while ((current_entry = hash_seq_search(&hash_seq)) != NULL)
+	{
+		BgwHandleStatus worker_status;
+
+		TerminateBackgroundWorker(current_entry->db_scheduler_handle);
+		worker_status = WaitForBackgroundWorkerShutdown(current_entry->db_scheduler_handle);
+		if (worker_status == BGWH_STOPPED)
+		{
+			tsbgw_total_workers_decrement();
+		}
+		else if (worker_status == BGWH_POSTMASTER_DIED) /* bailout */
+		{
+			proc_exit(1);
+		}
+
+	}
+}
+
+/*
+ * Drain queue, handle each message synchronously.
+ */
+static void
+launcher_handle_messages(HTAB *db_htab)
+{
+	tsbgwMessage *message;
+
+
+	for (message = tsbgw_message_receive(); message != NULL; message = tsbgw_message_receive())
+	{
+		tsbgwHashEntry *tsbgw_he;
+		PGPROC	   *sender;
+		bool		found = false;
+		VirtualTransactionId vxid;
+		bool		worker_registered;
+
+		sender = BackendPidGetProc(message->sender_pid);
+		if (sender != NULL)
+			GET_VXID_FROM_PGPROC(vxid, *sender);
+		else
+		{
+			ereport(LOG, (errmsg("timescalebgw cluster launcher received message from non-existent backend")));
+			continue;
+		}
+
+		switch (message->message_type)
+		{
+			case START:
+				ereport(LOG, (errmsg("in start condition")));
+
+				/*
+				 * we've just run check_for_stopped_schedulers, so we know if
+				 * we have found this in our hashtable that it is still alive
+				 * and we want to be idempotent, so don't need to do anything
+				 */
+				tsbgw_he = hash_search(db_htab, &message->db_oid, HASH_ENTER, &found);
+				if (!found)
+				{
+					if (!tsbgw_total_workers_increment(TSBGW_MAX_WORKERS_GUC_STANDIN))
+						ereport(ERROR, (errmsg("timescale background worker could not be started")));
+					worker_registered = register_tsbgw_entrypoint_for_db(tsbgw_he->db_oid, vxid, &tsbgw_he->db_scheduler_handle);
+					if (worker_registered)
+					{
+						pid_t		worker_pid;
+
+						WaitForBackgroundWorkerStartup(tsbgw_he->db_scheduler_handle, &worker_pid);
+						ereport(LOG, (errmsg("Worker started with PID %d", worker_pid)));
+					}
+				}
+				tsbgw_message_send_ack(message, true);
+				break;
+			case STOP:
+				ereport(LOG, (errmsg("in stop condition")));
+
+				tsbgw_he = hash_search(db_htab, &message->db_oid, HASH_REMOVE, &found);
+				if (found)
+				{
+					TerminateBackgroundWorker(tsbgw_he->db_scheduler_handle);
+					WaitForBackgroundWorkerShutdown(tsbgw_he->db_scheduler_handle);
+					tsbgw_total_workers_decrement();
+				}
+				tsbgw_message_send_ack(message, true);
+				break;
+
+			case RESTART:
+				ereport(LOG, (errmsg("in restart condition")));
```

Comment:
We decided against that because we want to maintain the worker's "slot" reserved in the number of workers, when you stop and start the num_workers will be decremented and then incremented, if you restart, the count simply shouldn't be affected. We don't want a race condition where some other db steals the scheduler of the other by requesting a worker at the wrong time. I can add a comment about this when I refactor to functions.